Add ModelView code (TreeStore and ListStore) from Gtk2Hs

hamishmack opened the pull request.

hamishmack committed:
Initial commit of ModelView code

See the haskell-gi branch in gtk2hs repo for history.

jplatte commented:
It would make sense to re-export all the new stuff from `Data.Gi.Gtk`, but there are some name conflicts between `Data.Gi.Gtk.ModelView.*` and `GI.Gtk.**`. I didn't look into everything in detail, but it seems like there is a difference in functionality.

If the ModelView functions from gtk2hs have no disadvantages over the equivalent generated functions (which I'd assume is the case), we should probably hide the generated functions.

Also, two nitpicks:
- `listStoreInsert` seems to be the gtk2hs equivalent of `listStoreInsertBefore`, it might make sense to rename it, and possibly add a new `listStoreInsertAfter` (and maybe even a new `listStoreInsert` without an argument for the position)
- Could you revert the formatting change of the cabal file? The main haskell-gi repo has the same formatting as we had before here, and I find it more readable as well. You should probably also add the authors of the gtk2hs modules (and yourself if you're not one of them) to the author field.

hamishmack committed:
Tidy up ModelView (removes the c2hs stuff)

hamishmack committed:
Rename treeDrag* to customDrag*

hamishmack committed:
Add seqStoreInsertBefore and seqStoreInsertAfter

hamishmack commented:
I have re-exported the modules from `Gtk`.

I renamed `TreeStore` -> `ForestStore` and `ListStore` -> `SeqStore` to avoid confusion with the Gtk types (the are designed to for the same purpose, but we can't use the Gtk Store functions on these `CustomStores`).

I renamed the `treeDrag*` functions in `CustomStore.hs` to `customDrag*`. 

There are still a two name clashes:
- `treeModelGetValue` adds a bit of type safety with `ColumnId`.
- The native `treeModelGetIter` is total, but gives an error when you use an empty path (something that is actually quite handy).

I added `seqStoreInsertBefore` and `seqStoreInsertAfter`.

I also fixed up the unintentional change to the cabal file format.  I was showing off Leksah's automatic `build-depends` feature to someone and forgot to revert.

garetxe commented:
Looks good to me. I must admit I did not check the patch in a lot of detail, but if it works and Hamish finds it useful I am all for including it. Jonas, if it looks good to you please go ahead and merge.

(By the way, apologies for the messages about the failing checks. I was setting up travis the other day for `haskell-gi-base`, and somehow I activated it for this repository too... Please ignore them for the moment.)

jplatte commented:
Yeah, I wanted to fast-forward, but whatever; I'll let GitHub have its merge commit. Not worth re-cloning the repository to get rid of that, as I seem to not have it in my current setup.

There's still the missing author information in the cabal file and the module export trick that breaks haddock that I noticed, but those really aren't what the PR is about. I'll fix them later.
